### PR TITLE
PWX-28162_pt3: Adding CSR auto-approval

### PR DESCRIPTION
**Pull request by zoxpx**
* adds new controller for CertificateSigningRequest (CSR) objects, which automatically approves TLS/SSL certificates requested by the OCI-Monitor
* functionality can be disabled by adding `portworx.io/disable-csr-approve: "true"` annotation, otherwise by default it will automatically approve PX-posted CSRs
* note that StorageCluster.Reconcile() needs to explicitly register the PX ClusterID with CSR-controller to handle only targetted (non-disabled) ClusterIDs

Signed-off-by: Zoran Rajic <zrajic@purestorage.com>

**What this PR does / why we need it**:

This is an improvement for the _"Automatic TLS setup"_ epic/work.

The TLS/SSL certificates currently need to be approved manually via `kubectl certificate approve px-node123`.
* as discussed w/ @piyush-nimbalkar, we are adding a dedicated controller that will monitor CertificateSigningRequest (CSR) objects, to automate this manual step, and automatically approve CSRs posted by the OCI-Monitor PODs
* note that after CSR is approved by PX-Operator, K8s will automatically attach the signed TLS/SSL cert to the CSR object
  - the PX installation watches CSRs, and will continue after TLS cert becomes available

**Which issue(s) this PR fixes** (optional)
Closes # PWX-28162  (part 3)

**Special notes for your reviewer**:

Additional dependent fixes:
* https://github.com/portworx/px-installer/pull/1525  -- independent fix for OCI-Monitor
* https://github.com/portworx/sched-ops/pull/294 -- `sched-ops` API update, **TODO** -- will need to vendor-in into the operator code

### CODE REVIEW

File: pkg/controller/csr/certificatesigningrequest.go
Comment (pure-jliao) [inline pkg/controller/csr/certificatesigningrequest.go:108]: maybe not good to stomp this

File: pkg/controller/storagecluster/storagecluster.go
Comment (pure-jliao) [inline pkg/controller/storagecluster/storagecluster.go:246]: I'm not sure if this is the best place to do this? probably after waiting for migration approval? idk

**Review by pure-jliao — COMMENTED** [review]
LGTM, will let others to take a look as well and approve

File: cmd/operator/operator.go
Comment (zoxpx) [reply cmd/operator/operator.go:12]: Oh, I'm surprised, you were actually maintaining the import-lists (and gaps) manually?

Most of our golang-code elsewhere is formatted according to `goimports` command -- which separates "golang internal" vs "everything else" -imports (like I did here).

File: drivers/storage/portworx/util/util.go
Comment (zoxpx) [reply drivers/storage/portworx/util/util.go:140]: This is supposed to be rarely used configuration -- where customer _wants_ automatic TLS, but _does not want_ operator to automatically approve the certs.
* in this case, they will have to manually approve the certs (i.e. `kubectl certificate approve XXX`), and the blocked PX-installs will continue.

File: pkg/controller/storagecluster/storagecluster.go
Comment (zoxpx) [reply pkg/controller/storagecluster/storagecluster.go:246]: I had to find a way how to bind the `StorageCluster`-controller updates, with CSR-controller
* note that StorageCluster YAML carries `disable-csr-approve` annotation that turns CSR-approval -controller on or off for that particular ClusterID
* so, we register the callback here, to configure the CSR-controller

File: pkg/controller/csr/certificatesigningrequest.go
Comment (zoxpx) [reply pkg/controller/csr/certificatesigningrequest.go:108]: Uh.. I'm not sure what you mean by this..   Could you please elaborate?

File: cmd/operator/operator.go
Comment (pure-jliao) [reply cmd/operator/operator.go:12]: IDE can sort the imports automatically, but here we want to separate the golang internal, portworx internal and everything else.

File: pkg/controller/csr/certificatesigningrequest.go
Comment (pure-jliao) [reply pkg/controller/csr/certificatesigningrequest.go:108]: do we want to print this in every reconcile loop every 20s? don't know if that's necessary.

File: cmd/operator/operator.go
Comment (zoxpx) [reply cmd/operator/operator.go:12]: Ok, if you guys keep re-sorting the lists manually, I'll revert my change.

File: pkg/controller/csr/certificatesigningrequest.go
Comment (zoxpx) [reply pkg/controller/csr/certificatesigningrequest.go:108]: Oh, I thought controller's `Reconcile()` triggers on K8s object's _changes_  (didn't realize it triggers on schedule, every 20 seconds).   If this is the case, what's the advantage of using K8s controllers in the fist place  (seems like cronjobs would do a similar job, and object watches would be more efficient?)

Other than this..
* the StorageCluster's `Reconcile()` will also drop an INFO-log every time it triggers
* also, since CSR-controller is a brand new functionality, I think it would be OK to have more verbose logging until we have better confidence the new code works w/o issues...

File: pkg/controller/csr/certificatesigningrequest.go
Comment (piyush-nimbalkar) [inline pkg/controller/csr/certificatesigningrequest.go:150]: nit: We can check `len(labs) == 0`. That way we check for both `nil` and empty map.

File: pkg/controller/csr/certificatesigningrequest.go
Comment (piyush-nimbalkar) [inline pkg/controller/csr/certificatesigningrequest.go:33]: To avoid defining constants in multiple places, can we make the existing constants public and use it in this controller?
Here is the reference to them: https://github.com/libopenstorage/operator/blob/master/drivers/storage/portworx/util/util.go#L604

File: pkg/controller/csr/certificatesigningrequest.go
Comment (piyush-nimbalkar) [inline pkg/controller/csr/certificatesigningrequest.go:218]: nit: Should we check if the condition of that type already exists and edit that instead of directly appending it?

File: pkg/controller/storagecluster/storagecluster.go
Comment (piyush-nimbalkar) [inline pkg/controller/storagecluster/storagecluster.go:246]: Can we also unregister on StorageCluster delete so we no longer approve any further/pending CSRs with that ID? 
We can do that in the `deleteStorageCluster` function -  https://github.com/libopenstorage/operator/blob/ff3634adc3ffd93ad17dc7ba0e92d55682fc57a1/pkg/controller/storagecluster/storagecluster.go#L649

**Review by piyush-nimbalkar — APPROVED** [review]
Looks good, @zoxpx. Just a comment about unregister, rest are just nits.

File: pkg/controller/csr/certificatesigningrequest.go
Comment (zoxpx) [reply pkg/controller/csr/certificatesigningrequest.go:56]: You are correct -- thanks for pointing it out
(although I don't think many people will run the patch-level 0 `k8s-1.19.0` version, but it's still a bug)

File: pkg/controller/csr/certificatesigningrequest.go
Comment (zoxpx) [reply pkg/controller/csr/certificatesigningrequest.go:218]: Note -- we don't actually call the `appendCondition()` "unconditionally" --
* if CSR has been approved/denied, we'll exit `Reconcile()`

ref from `Reconcile():137`
```go
	if approved, denied := c.getCertApprovalCondition(&csr.Status); approved || denied {
		log.WithField("approved", approved).WithField("denied", denied).
			Debugf("CSR already approved/denied")
		return ret, nil
	}
```

**Comment by zoxpx** [comment]
**NOTE**: Piyush'es feedback addressed with [df4461c](https://github.com/libopenstorage/operator/pull/874/commits/df4461c0f76a463e891216d251b5e8093489219e) commit

**Comment by zoxpx** [comment]
Thanks for the review guys -- merging the PR.

**Pull request merged**
